RT11xx SOC level power management

danieldegrasse opened the pull request:
Enables SOC level power management for the RT11xx series, enabled when `CONFIG_PM=y`. When power management is enabled, the SOC will transition to STOP mode during longer sleep periods, and will additionally reduce its core voltage and disable peripheral clocks where possible.

danieldegrasse committed:
soc: rt11xx: Enable power management for RT11xx series SOCs

Enables basic power management for the RT11xx series SOCs. The following
power saving measures are currently implemented:

- system reduces core voltage during sleep
- core clock can be gated in deep sleep

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
soc: mimxrt1160: Add power state definitions for RT1160

Adds power state definitions for RT1160. Min residency times are loosely
derived from transition times into similar power states in the datasheet

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: timer: mcux_gpt: Enable GPT timer to run in doze mode

GPT timer must continue running in low power modes, as it is the system
wakeup source. Set configuration to ensure peripheral will not stop
running in low power modes.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
soc: imx_rt: Enable IMX.RT General Power Controller Driver

GPC driver is used for power management operations on RT series SoCs.
Enable inclusion of the driver into the build.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

ceolin commented:
I not familiar with this target but I put some comments related with the pm subsystem for you review.

ceolin commented on boards/arm/mimxrt1160_evk/mimxrt1160_evk.dtsi:
```diff
@@ -33,6 +34,26 @@
+			min-residency-us = <10>;
+		};
+
+		suspend: set_point_10_stop {
```
s/set_point_10_stop/set_point_10_suspend ?

danieldegrasse replied:
Good point. Renamed it.

ceolin commented on boards/arm/mimxrt1170_evk/mimxrt1170_evk.dtsi:
```diff
@@ -33,6 +35,25 @@
+			min-residency-us = <10>;
+		};
+
+		suspend: set_point_10_stop {
```
see previous comment about node name

danieldegrasse replied:
Fixed in latest push, thanks

ceolin commented on soc/arm/nxp_imx/rt/power_rt11xx.c:
```diff
@@ -0,0 +1,337 @@
+	uint32_t tmp;
+
+#ifdef CONFIG_CPU_CORTEX_M7
+	uint8_t mapping[17][17] = CPU0_COMPATIBLE_SP_TABLE;
```
17 is kinda of crypt, may a constant more descriptive for this value ?

danieldegrasse replied:
Fixed this to use a constant

ceolin commented on soc/arm/nxp_imx/rt/power_rt11xx.c:
```diff
@@ -0,0 +1,337 @@
+	 * waking up from any interrupt which priority is not 0. Work around the
+	 * issue and disable interrupts using PRIMASK register as recommended
+	 * by ARM.
+	 */
```
The thing is that we can't be interrupted during the sleep process in the idle
thread until we call arch specific code. Then its is up to the target set up
which interruption should wake up the system.

danieldegrasse replied:
So to clarify, the kernel is disabling interrupts to avoid the sleep process being interrupted, but once SOC specific power management code is called the SOC is free to enable interrupts again? If so, I'm assuming that I should just change the comment here to reflect the correct reason that the BASEPRI register needs to be set.

ceolin commented on soc/arm/nxp_imx/rt/power_rt11xx.c:
```diff
@@ -0,0 +1,337 @@
+		GPC_CM_RequestRunModeSetPointTransition(GPC_CPU_MODE_CTRL,
+							current_set_point);
+		cpu_mode_transition(gpc_mode, true);
+	} else if (current_set_point != set_point) {
```
This should not be necessary, the subsystem only calls this function from the
active state. In other words, will never be called from a sleep state.

danieldegrasse replied:
Thanks for the heads up- I'll remove this block

dleach02 commented on boards/arm/mimxrt1160_evk/mimxrt1160_evk.dtsi:
```diff
@@ -33,6 +34,26 @@
+			/*  suspend corresponds to set point 10 for RT1170 */
+			compatible = "zephyr,power-state";
+			power-state-name="suspend-to-idle";
+			substate-id = <IMX_SPC_SET_POINT_10_SUSPEND>;
```
Curious why set point 10?

danieldegrasse replied:
Set point 10 seemed to have a good combination of lower wakeup time and current consumption. The RT1170 [datasheet](https://www.nxp.com/docs/en/data-sheet/IMXRT1170CEC.pdf) states set point 10 has a typical wakeup time of ~6.74ms (pg 35), and disables all clock sources and PLLs except the 32KHz oscillator (which the GPT runs from). The `imx_spc.h` file defines all set point configurations, so we can switch to a different set point if we need to (as can users when building custom boards)

dleach02 replied:
Probably useful to have a comment block somewhere in this code to 1) reference the datasheet and 2) give your rational... That would give future folks some insight in the selection process you went through and provide them with guidance to select (or experiment) with different setpoints potentially in an application dts overlay

hakehuang commented:
@danieldegrasse  can I run the tests\subsys\pm against this PR?

danieldegrasse commented:
@hakehuang  If you can, that would be great. This PR is dependent on #41245, so that change will either have to be made manually to run CI, or we can wait until it merges and I'll rebase

hakehuang commented:
@danieldegrasse , the kernel context test found there is 1 cycle diff 

START - test_kernel_cpu_idle
Assertion failed at WEST_TOPDIR/zephyr/tests/kernel/context/src/main.c:352: _test_kernel_cpu_idle: (tms2 < tms is true)
Bad ms value computed,got 5479 which is less than 5480
FAIL - test_kernel_cpu_idle in 0.18 seconds

```
INFO    - Added initial list of jobs to queue
INFO    - 962/343 mimxrt1170_evk_cm7        tests/kernel/context/kernel.common                 FAILED Failed (device 7.959s)
INFO    - /home/shared/disk/zephyr_project/zephyr_test/zephyr/twister-out/mimxrt1170_evk_cm7/tests/kernel/context/kernel.common/handler.log
ERROR   - *** Booting Zephyr OS build v2.7.99-2480-g7401a46e0f40  ***
Running test suite context
===================================================================
START - test_kernel_interrupts
PASS - test_kernel_interrupts in 0.1 seconds
===================================================================
START - test_kernel_ctx_thread
Testing k_current_get() from an ISR and thread
Testing k_is_in_isr() from an ISR
Testing k_is_in_isr() from a preemptible thread
PASS - test_kernel_ctx_thread in 0.12 seconds
===================================================================
START - test_busy_wait
Thread busy waiting for 20000 usecs
Thread busy waiting completed
Thread busy waiting for 20000 usecs (irqs locked)
Thread busy waiting completed (irqs locked)
PASS - test_busy_wait in 0.54 seconds
===================================================================
START - test_k_sleep
thread sleeping for 50 milliseconds
thread back from sleep
Testing k_thread_create() without cancellation
thread (q order: 2, t/o: 500) is running
got thread (q order: 2, t/o: 500) as expected
thread (q order: 3, t/o: 750) is running
got thread (q order: 3, t/o: 750) as expected
thread (q order: 0, t/o: 1000) is running
got thread (q order: 0, t/o: 1000) as expected
thread (q order: 6, t/o: 1250) is running
got thread (q order: 6, t/o: 1250) as expected
thread (q order: 1, t/o: 1500) is running
got thread (q order: 1, t/o: 1500) as expected
thread (q order: 4, t/o: 1750) is running
got thread (q order: 4, t/o: 1750) as expected
thread (q order: 5, t/o: 2000) is running
got thread (q order: 5, t/o: 2000) as expected
Testing k_thread_create() with cancellations
cancelling [q order: 0, t/o: 1000, t/o order: 0]
thread (q order: 3, t/o: 750) is running
got (q order: 3, t/o: 750, t/o order 1) as expected
thread (q order: 0, t/o: 1000) is running
got (q order: 0, t/o: 1000, t/o order 2) as expected
cancelling [q order: 3, t/o: 750, t/o order: 3]
cancelling [q order: 4, t/o: 1750, t/o order: 4]
thread (q order: 4, t/o: 1750) is running
got (q order: 4, t/o: 1750, t/o order 5) as expected
cancelling [q order: 6, t/o: 1250, t/o order: 6]
PASS - test_k_sleep in 5.332 seconds
===================================================================
START - test_kernel_cpu_idle_atomic
SKIP - test_kernel_cpu_idle_atomic in 0.1 seconds
===================================================================
START - test_kernel_cpu_idle
Assertion failed at WEST_TOPDIR/zephyr/tests/kernel/context/src/main.c:352: _test_kernel_cpu_idle: (tms2 < tms is true)
Bad ms value computed,got 5479 which is less than 5480
FAIL - test_kernel_cpu_idle in 0.18 seconds
===================================================================
START - test_k_yield
PASS - test_k_yield in 0.1 seconds
===================================================================
START - test_kernel_thread
PASS - test_kernel_thread in 0.1 seconds
===================================================================
START - test_kernel_timer_interrupts
SKIP - test_kernel_timer_interrupts in 0.0 seconds
===================================================================
Test suite context failed.
===================================================================
PROJECT EXECUTION FAILED


```

danieldegrasse commented:
@hakehuang That test failure looks to be an issue with the assumptions the test makes (one of which the GPT timer driver violates). I opened an issue (#41347), and PR (#41348) to fix it. Since the test fails on main, I don't think that this issue is related to or triggered by the PM changes, it is a result of using the GPT timer.

hakehuang reviewed (approved):
all RT1170 regression test PASS

carlescufi commented:
@gmarull could you take a look please?

gmarull commented on boards/arm/mimxrt1160_evk/mimxrt1160_evk.dtsi:
```diff
@@ -33,6 +34,35 @@
 	};
+
+
+	power-states {
+		/*
+		 * RT1160 Power states are managed with set points (see page 30-33)
+		 * of the datasheet). These set points correspond to various power
+		 * savings, and associated transition (residency) times.
+		 *
+		 * Set points 1 and 10 were chosen as sane defaults to offer
+		 * limited power savings and quick transitions when entering idle for
+		 * short periods, and better power savings with longer transition
+		 * times for long idle periods
+		 */
+		idle: set_point_1_wait {
```
note: wouldn't it be better that SoC dtsi files provide some defaults that all boards inherit by default? If needed board can always override (creating extra set-points). I don't know enough this platform to judge, though.

danieldegrasse replied:
Sorry for the delay in response- it's probably best to set defaults at the SOC level. I updated the PR to do that, rather than define power states at the board level.

cfriedt commented:
@dleach02 - does this look OK?

danieldegrasse committed:
soc: mimxrt1170: Add power state defintions for RT1170

Adds power state definitions for RT1170. Min residency times are loosely
derived from transition times into similar power states in the datasheet

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>
